File: pbs/qsub.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Simple wrapper of qsub command of PBS system
"""
import argparse
import subprocess
import os


def make_script(args, rest):
    executable = os.path.abspath(args.executable)
    assert os.access(executable, os.X_OK)
    lines = [
        '#!/bin/sh',
        'cd $PBS_O_WORKDIR',
        ' '.join([executable] + rest)
    ]
    return '\n'.join(lines) + '\n'


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('-n', '--dry-run', action='store_true')
    parser.add_argument('-N', '--jobname')
    parser.add_argument('-j', '--ncpus', default=1, type=int)
    parser.add_argument('-m', '--mem', default=1, type=int)
    parser.add_argument('executable')
    (args, rest) = parser.parse_known_args()
    jobname = args.jobname or os.path.basename(args.executable)
    resource = 'select=1:ncpus={0.ncpus}:mem={0.mem}gb'.format(args)
    command = f'qsub -j oe -N {jobname} -l {resource}'
    if args.executable.endswith('.R'):
        qsub_R = os.path.join(os.path.dirname(__file__), 'qsub.R')
        command += f' -v NCPUS={args.ncpus},INFILE={args.executable} {qsub_R}'
        input = None
    else:
        script = make_script(args, rest)
        input = script.encode()
        print(script)
    print(command)
    if not args.dry_run:
        subprocess.run(command, input=input, shell=True)


if __name__ == '__main__':
    main()

File: pbs/qsub.R
#!/usr/bin/env Rscript
#PBS -N qsub.R
#PBS -j oe
#PBS -l select=1:ncpus=24:mem=24gb
#
# qsub.py -j24 -m24 -N jobname path/to/script.R
#
# qsub -v NCPUS=24,INFILE=sleep.R -N jobname qsub.R
# NCPUS=3 INFILE=envvars.R qsub.R
#
stopifnot(!is.na(Sys.getenv("INFILE", NA)))
script = Sys.getenv("INFILE", NA)
PBS_O_WORKDIR = Sys.getenv("PBS_O_WORKDIR", NA)
if (!is.na(PBS_O_WORKDIR)) {
  setwd(PBS_O_WORKDIR)
  options(mc.cores = Sys.getenv("NCPUS", 24L))
} else {
  # for testing
  options(mc.cores = Sys.getenv("NCPUS", 2L))
}
message("mc.cores:", getOption("mc.cores"))
message("getwd():", getwd())
message("source('", script, "') ##########")
source(script, echo = FALSE)
